File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
from collections import deque
r = sys.stdin.readline
sys.setrecursionlimit(10**6)

def solution():
    while True:
        q = deque()
        line = r().rstrip()
        if line == '.': break
        for c in line:
            if c == '[' or c == '(':
                q.append(c)
            elif c == ']':
                if q and q[-1] == '[':
                    q.pop()
                else: 
                    q.append(c)
                    break
            elif c == ')':
                if q and q[-1] == '(':
                    q.pop()
                else: 
                    q.append(c)
                    break
            else: continue
        
        if len(q) == 0: print('yes')
        else: print('no')
    
solution()

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        while(true) {
            String[] line = br.readLine().split("");
            if (line[0].equals(".")) break;
            
            Stack<String> stk = new Stack<>();
            
            for(String c : line) {
                switch (c) {
                    case "(":
                        stk.add(c);
                        break;
                    case "[":
                        stk.add(c);
                        break;
                    case ")":
                        if (!stk.isEmpty() && stk.peek().equals("(")) stk.pop();
                        else stk.add(c);
                        break;
                    case "]":
                        if (!stk.isEmpty() && stk.peek().equals("[")) stk.pop();
                        else stk.add(c);
                        break;
                }
            }
            
            if (stk.size() == 0) System.out.println("yes");
            else System.out.println("no");
        }
    }
}